flycheck.el

```diff
@@ -7941,6 +7960,26 @@ Requires Go 1.6 or newer. See URL
   ((warning line-start (file-name) ":" line ":" column ": " (message) line-end))
   :modes go-mode)
 
+(flycheck-define-checker go-staticcheck
+  "A Go checker that performs static analysis and linting using
+the `staticcheck' command. `staticcheck' is the successor to
+`megacheck'; while the latter isn't fully deprecated yet, it's
+recommended to migrate to `staticcheck'.
+
+`staticcheck' is explicitly fully compatible with \"the last two
+versions of go\". `staticheck' can target earlier versions (with
+limited features) if `flycheck-go-version' is set. See URL
+`https://staticcheck.io/'."
+  :command ("staticcheck"
+            (option-list "-tags" flycheck-go-build-tags concat)
+            (option "-go" flycheck-go-version)
+            ;; Run in current directory to make staticcheck aware of symbols
+            ;; declared in other files.
+            ".")
```

Comment:
So if you pass the current file to to staticcheck, it will not look at other imported files?

Reply:
Mmmm, that's a great question. `staticcheck` evolved out of `megacheck`, which (to my understanding) was persnickity about these things. Likely, `staticcheck` isn't, and I'm being overly cautious/cargo-cult-y. Will confirm how this works.

Reply:
So: checking in on this has turned up two things. 

First, to answer your question: if you just give it a file, `staticcheck` will report errors, but not warnings, for other files imported by the file in question. This seems fine to me, and updating the checker to not specify `"."` is easy enough. However, I also remark that this checker, as I've written it, does not exhibit a behaviour I see in other checkers where errors in other files are  reported in the file currently under check. I think it would be correct for this checker to have that behaviour; I'll go try and learn how to do that, but if you have the time to point me in a direction, it would be appreciated!

Second thing I found: `staticcheck` _does_ have a notion of "severity", but it only reports severity if you format its output as JSON. I think it would make this checker dramatically more useful if severity were correctly reported, so I'm going to go see if I can figure out how to parse the JSON output. I don't know that that needs to block this PR; setting the default level of the checker to `warning` or `error` seems like an OK version one to me, if it works for you. But, maybe it's quite easy to ingest JSON in flycheck? I have no idea.

Reply:
Ah so! Three birds, one stone: it turns out, flycheck makes parsing JSON errors extremely easy! (I should have guessed that this would be true -- flycheck has been a lovely project to contribute to.) Modifying this checker to 1) not use "." and 2) to correctly parse JSON output has _also_ resulted in flycheck reporting errors in other files. Fantastic!

I'm adding a buttercup test for the new JSON parser. I've also realized I managed to never commit or push the test files required for the go-staticheck tests I wrote -- so that's embarrassing, but also an easy fix. Will get all the changes pushed very soon here :+1:

Reply:
Thanks for looking into this.  I left another question in the comments below.